### ISSUE TITLE
YMT Limits (Request)

### ISSUE THREAD

**Issue by LivedBackwardss** [citizenfx/fivem#2938]
### What happened?

With each new version, the .ymt limit is reduced, which forces us to adjust the clothing and make it compatible with the updated limits. Similar to how we can adjust the pool size, would it be possible to add a feature that allows us to increase the limit to a maximum of 14 or something similar? 

> In game version b1604 (mpchristmas2018) we had 14 additional free .ymt slots to be used  
> In game version b1737 (mpvinewood) we had 12 additional free .ymt slots to be used  
> In game version b1868 (mpheist3) we had 10 additional free .ymt slots to be used  
> In game version b2060 (mpsum) we had 8 additional free .ymt slots to be used  
> In game version b2189 (mpheist4) we had 6 additional free .ymt slots to be used  
> In game version b2372 (mptuner) we had 4 additional free .ymt slots to be used  
> In game version b2545 (mpsecurity) we have 2 additional free .ymt slots to be used  
> In game version b2612 (mpg9ec), R* seems to have increased the YMT limit, we have about 20 additional free .ymt slots to be used  
> In game version b2699 (mpsum2), we have about 16 additional free .ymt slots to be used  
> In game version b2802/b2824/b2845 (mpchristmas3), we have about 14 additional free .ymt slots to be used  
> In game version b2944 (mp2023_01), we have about 12 additional free .ymt slots to be used  
> In game version b3095 (mp2023_02), we have about 10 additional free .ymt slots to be used  
> In latest game version b3258 (mp2024_01), we have about 8 additional free .ymt slots to be used  

For example, `increase_pool_size "YmtLimit" 14`. This limit could be achievable because the Emergency Uniform Pack (9.8) itself takes up 2 .ymt slots. Then, you could use the ONX faces, which take up another 2 .ymt slots, bringing you to 4/8 for the build, leaving only four .ymt slots remaining. It doesn’t make sense that the .ymt limit gets reduced with every patch, forcing servers to adjust their content each time the game updates. This creates unnecessary work for server owners and developers.

### Expected result

The clothing should load regardless of how many slots available. It doesn't make sense that from 1604 to 3258 that we go from 14 to 8.

### Reproduction steps

1. Get a clothing pack that breaks the limit.
2. Some other clothing won't load or faces.

### Importancy

Slight inconvenience

### Area(s)

FiveM, FXServer

### Specific version(s)

FiveM 3258 / Server 11012 Windows

### Additional information

FiveM Game Build: 3258
Artifact: 11012 (Windows)

**Comment by Keep874** [comment]
Agree, this is an inconvenience to server owners having to adjust their limits per game update.

**Comment by vipexv** [comment]
Would be cool if possible :o

**Comment by LivedBackwardss** [comment]
> let's be honest, do you really want to redo your EUP every game update because of new GTA clothing in the update. Plus you need to have your players fix all of their outfits. I say pick a gamebuild that works for you and stick with it. Then, get on these clothing sellers to stop encrypting so that you can add their new cool clothes to your custom ymt files. let's be honest, most of these clothing sellers are just using stock GTA ydd files anyway.

Absolutely not, I don't want to change my clothing packs but I do want the latest vehicles and game features, so it has to be done. It makes no sense how the limit goes down every update. We should be able to set it like a pool size like the other stuff.

**Comment by ook3D** [comment]
> 
>  It makes no sense how the limit goes down every update.

It does though the online update adds new stuff each update how does that not make sense

**Comment by Rm1998** [comment]
> let's be honest, do you really want to redo your EUP every game update because of new GTA clothing in the update. Plus you need to have your players fix all of their outfits. I say pick a gamebuild that works for you and stick with it. Then, get on these clothing sellers to stop encrypting so that you can add their new cool clothes to your custom ymt files. let's be honest, most of these clothing sellers are just using stock GTA ydd files anyway.

You already can add encrypted clothes to your custom .ymt files. As long as the encrypted .ydd's are in the correct folder it doesn't matter what ymt you attach them to.

**Comment by LivedBackwardss** [comment]
Still would be nice to have this as a pool size, every update introduced we go down slots. Please allow us to set the amount of ymts or something.

> In latest game version b3407 (mp2024_02), we have about 6 additional free .ymt slots to be used

**Comment by JHansen2000** [comment]
I agree with this, the ability to override the number of free .ymt slots is critical, especially as we approach a point in which it becomes increasingly unlikely that new DLCs will increase the number of slots.

**Comment by Gltch3d** [comment]
I don't think this is a good idea until this issue (https://github.com/citizenfx/fivem/issues/2984) is addressed as any further increases of limits would only allow servers to reach the overall pgRawStreamer limit

**Comment by JHansen2000** [comment]
> I don't think this is a good idea until this issue (https://github.com/citizenfx/fivem/issues/2984) is addressed as any further increases of limits would only allow servers to reach the overall pgRawStreamer limit faster. Just combine your clothing assets under a existing ymt's. Even if it's escrowed you can still stream the files in another resource and have them added to a ymt. Just need to make sure the file's are named appropriately 

I disagree. As of the latest DLC we only have 6 YMTs to use, and this number will only get smaller with any future releases. The solution you've described is hacky at best. Increasing the pgRawsStreamer limit, while somewhat beneficial, would be a massive overhaul akin to the release of OneSync (which is what increased to the current 65535 limit) and may actually be detrimental to clients. This massive overhaul should not be a blocker for other more achievable improvements.

Truthfully, in my opinion, hitting the pgRawStreamer limit is a reflection of a poorly optimized server. If anything, servers being affected by this should write better scripts and/or apply pressure to their script providers to do the same. Increasing the limit would only encourage bad coding practices.

**Comment by Gltch3d** [comment]
> I disagree. As of the latest DLC we only have 6 YMTs to use, and this number will only get smaller with any future releases. The solution you've described is hacky at best. Increasing the pgRawsStreamer limit, while somewhat beneficial, would be a massive overhaul akin to the release of OneSync (which is what increased to the current 65535 limit) and may actually be detrimental to clients. This massive overhaul should not be a blocker for other more achievable improvements.
> 
> Truthfully, in my opinion, hitting the pgRawStreamer limit is a reflection of a poorly optimized server. If anything, servers being affected by this should write better scripts and/or apply pressure to their script providers to do the same. Increasing the limit would only encourage bad coding practices.

pgRawsStreamer has nothing to do with coding practices. It is the overall limit of streamed assets, ydd, and ytd's included. More ymt's would allow for more of these files to be streamed which would also bring a server closer to this limit. Better scripting practices wouldn't prevent this when they're calculated on startup. By the way, it's very easy for a server that has been actively adding new content for 2 or more years to reach this limit relatively fast (this includes clothing, vehicles, map edits). As with each DLC update it not only reduces the amount of additional YMT's you can stream but it also reduces the amount of assets you can stream.

**Comment by DaniGP17** [comment]
Can someone provide crash dump?

**Comment by LivedBackwardss** [comment]
> Can someone provide crash dump?

Hey, there's some details about the crashes on this page about the limitations: https://docs.gta.clothing/game-mechanics/game-limits-and-crashes#ymt-limit-in-detail and there's a thread on the forums at https://forum.cfx.re/t/setplayermodel-crash-when-you-have-too-much-ymts-assigned-to-a-freemode-ped-model/1934063 with some crashes/debugs. Most of the time the items do not load but it doesn't make sense that the limitations decrease each build.

**Comment by Chip-W** [comment]
What most here don't understand about why the ymt limit goes down is the reason for it is because Rockstar is using part of the limit with their DLC clothing.

With every DLC, rockstar is taking up at least 2 .ymt's for new clothes (1 male, 1 female).  Sometimes it's more.  There are only so many .ymt slots available overall, so everytime they release a DLC, they are taking up YOUR ymt slots.  This is why they go down and there isn't a lot Cfx can do about it.  The last time it was raised was because ROCKSTAR was running out of slots and needed more for future DLC's.  So ROCKSTAR wrote in the increase, not Cfx.

**Comment by LivedBackwardss** [comment]
> What most here don't understand about why the ymt limit goes down is the reason for it is because Rockstar is using part of the limit with their DLC clothing.
> 
> With every DLC, rockstar is taking up at least 2 .ymt's for new clothes (1 male, 1 female). Sometimes it's more. There are only so many .ymt slots available overall, so everytime they release a DLC, they are taking up YOUR ymt slots. This is why they go down and there isn't a lot Cfx can do about it. The last time it was raised was because ROCKSTAR was running out of slots and needed more for future DLC's. So ROCKSTAR wrote in the increase, not Cfx.

So, there's no way to get more?

**Comment by SpellLivedBackwards** [comment]
Any updates? The pull request was closed seven months ago, and we still haven’t received a response from the team. Many users are eager to upgrade to the latest version of the game build but are unable to do so due to these limitations.

**Comment by Chip-W** [comment]
> This would be a huge help, our server uses 1 ymt for civs and 1 for leo, we have a dog ped so that takes about another and then roxwood, we have had to turn most of our addon peds and animals to only being turned on by request because the server just cant run everything anymore. a increase to this would be a huge help and not limiting. we have hit the point we can not update game builds because if we do the content creators that fund our community couldnt use the server for their videos because we couldnt stream their peds anymore

Actually, if your pack has both male and female, then it takes TWO .ymt slots.  Then, if there is a creaturemetadata, that is a 3rd .ymt slot for just one pack.  As far as Roxwood goes, I just removed the clothing resource from it.  I don't use their uniforms anyway.

**Comment by aymannajim** [comment]
It’s been a month since the last reply. We really need this addressed! This feature is very important for taking full advantage of the latest game builds while keeping the same clothing pack.

**Comment by Legacy-TacticalGamingInteractive** [comment]
Yup we are stuck on 3258 too but what can you do really. Rockstar is probably not going to update the ymt limit again. We can only hope... so we can purchase more clothing from Tebex stores. Until then you can probably consider a lot of clothing development to slow down to a halt eventually. Because anyone making custom clothing designs is not going to want to immediately have all of their stuff unescrowed - but then they will lose sales if they do have it escrowed. Its unfortunate.

**Comment by Legacy-TacticalGamingInteractive** [comment]
> > I don't think this is a good idea until this issue ([#2984](https://github.com/citizenfx/fivem/issues/2984)) is addressed as any further increases of limits would only allow servers to reach the overall pgRawStreamer limit faster. Just combine your clothing assets under a existing ymt's. Even if it's escrowed you can still stream the files in another resource and have them added to a ymt. Just need to make sure the file's are named appropriately
> 
> I disagree. As of the latest DLC we only have 6 YMTs to use, and this number will only get smaller with any future releases. The solution you've described is hacky at best. Increasing the pgRawsStreamer limit, while somewhat beneficial, would be a massive overhaul akin to the release of OneSync (which is what increased to the current 65535 limit) and may actually be detrimental to clients. This massive overhaul should not be a blocker for other more achievable improvements.
> 
> Truthfully, in my opinion, hitting the pgRawStreamer limit is a reflection of a poorly optimized server. If anything, servers being affected by this should write better scripts and/or apply pressure to their script providers to do the same. Increasing the limit would only encourage bad coding practices.

Tell it to the mappers and dev teams like Roxwood who don't know how to properly use manifests so instead they overwrite (stream) thousands of vanilla props for no reason and rename them/stream them on servers. I mean ffs one of them is a farm trailer. They changed it from yellow to green... wtf was the point. Lets not get started on 1000000 lodDist lol. Some servers simply buy a lot of stuff as well because they want to give content to their players. The game already runs at 65-85% archetypes streamed vanilla with no mods in many places on the map. Add a couple of larger map additions and you hit that in no time. Mile High Club is like 984 assets. Roxwood around 3300 or something. There are countless times we see MLO's come with duplicate defined archetypes in their YTYP of vanilla props they simply needed to reference with a manifest for no reason at all. Adding to the total registered archtypes. There are some servers with a lot of assets that are still optimized but you would be right to say many are not at all especially by novice owners and "dev" teams just dragging and dropping things. Anyways this is slightly off topic.

On the topic of EUP... well... thats a whole other issue. Indeed you have many people making vests or shirts and instead of utilizing multiple textures per model they duplicate it and put one texture on each model. There are ways to certainly condense EUP and things but we can all only do so much.

Anyways its a moot point probably nothing will happen unless someone tries to convince Rockstar themselves to update GTAV with the change... lol. But fun to discuss. 3258 it is.

**Comment by lLukDreams** [comment]
@prikolium-cfx what is status on this?

**Comment by Legacy-TacticalGamingInteractive** [comment]
Perhaps they dont have access to that code or ability to change the limit themselves without the GTAV team itself doing it first. Whether or not they are with Rockstar. Who knows. Maybe one of them can shed some light on the possibility.

**Comment by ook3D** [comment]
> Perhaps they dont have access to that code or ability to change the limit themselves without the GTAV team itself doing it first. Whether or not they are with Rockstar. Who knows. Maybe one of them can shed some light on the possibility.

FiveM was built without having access to the code, I don't see that as a valid reason, no point in speculating

**Comment by Gogsi** [comment]
It's technically possible to increase the limited with runtime patching but it's really annoying because of the sheer amount of arrays that need to be patched.  #3444 tried to increase the limit but only on build 1604 (which would have the least code that needs to be patched) and it didn't fully work, resulting in other crashes. Some of the work done in #3477 could simplify the code needed to patch a function's stack (via the `IncreaseFunctionStack` utility) but that code isn't merged yet.

**Comment by DaniGP17** [comment]
Yes, although I found the possible cause of some crashes, manually resizing all the arrays was very tedious and time-consuming.  However, I noticed that Ehbw implemented something like an "automatic" stack resizing feature in one of his PRs, which might be used in the future to try and increase this limit again.

**Comment by SpellLivedBackwards** [comment]
It feels like issues related to limitations never get a response, yet requests for pool size increases are addressed. This particular issue has been open for nearly a year, and it's surprising that we haven’t heard anything from the team. With each new game build, we’re forced to reduce the amount of clothing, and with store content being escrowed, it’s becoming nearly impossible to manage. It’s similar to the random texture loss issue, but even more frustrating—especially since many servers want to adopt the latest build but are blocked by YMT limits.

> Yes, although I found the possible cause of some crashes, manually resizing all the arrays was very tedious and time-consuming. However, I noticed that Ehbw implemented something like an "automatic" stack resizing feature in one of his PRs, which might be used in the future to try and increase this limit again.

Isn’t that something for Red Dead Redemption II, or does it also apply to GTA V?

**Comment by Legacy-TacticalGamingInteractive** [comment]
> It feels like issues related to limitations never get a response, yet requests for pool size increases are addressed. This particular issue has been open for nearly a year, and it's surprising that we haven’t heard anything from the team. With each new game build, we’re forced to reduce the amount of clothing, and with store content being escrowed, it’s becoming nearly impossible to manage. It’s similar to the random texture loss issue, but even more frustrating—especially since many servers want to adopt the latest build but are blocked by YMT limits.

Maybe you're right. I had an issue open  which I don't think I ever heard back about which was for Carcols Siren ID changing them from 16 bit to 32 bit integer (to go beyond 255 to 65535) either. Even though we did it with clothing props. But that just makes me wonder if its just too much work or something.

But we must not forget... i will say that I only see a handful of cfx doing stuff on the repo though because no doubt they are busy af. Like @prikolium-cfx is certainly one of the most active along with @dalekenium-cfx @nihonium-cfx @Nobelium-cfx @neptunium-cfx. It seems to be a very small team AND some of them might not even be around anymore. So you can see why it can be overwhelming to take on bigger issues. Why its also best if the community can provide a PR to get the ball rolling I'm sure but even then.

**Comment by SpellLivedBackwards** [comment]
With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, @DaniGP17's pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.

**Comment by Identity-labs** [comment]
Try to find a workaround, don't wait anything from Cfx, it been many years like that.

**Comment by Scullyy** [comment]
> With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, [@DaniGP17](https://github.com/DaniGP17)'s pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.

The pool size increases that were made are much easier than what's being requested here. It's difficult to raise this limit without causing other problems. Even on the latest game build you can already have hundreds of clothing items, you don't need to have a different ymt for every single clothing piece you purchase. You just need the clothing within that resource but you can place them all inside a single ymt.

**Comment by SpellLivedBackwards** [comment]
> Try to find a workaround, don't wait anything from Cfx, it been many years like that.

The only workaround is to reduce the number of .ymt files, which is clearly not feasible due to certain assets being locked in escrow. For example, ONX faces already use up 2 .ymts — one for male and one for female. Then, the Emergency Uniform Pack takes up another 2 .ymts. That only leaves a single slot available for the latest Money Fronts build, given the 5 .ymt file limit.


> > With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, [@DaniGP17](https://github.com/DaniGP17)'s pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.
> 
> The pool size increases that were made are much easier than what's being requested here. It's difficult to raise this limit without causing other problems. Even on the latest game build you can already have hundreds of clothing items, you don't need to have a different ymt for every single clothing piece you purchase. You just need the clothing within that resource but you can place them all inside a single ymt.

So you're suggesting we merge all the .ymt files into one? Isn't there a limit to how many clothing items can be included per pack — like 128 per clothing category in a pack, for example? Plus, doesn't asset escrow come into play here? I don't think merging is a realistic option, especially when people are asset-escrowing their clothing packs.

**Comment by ghost** [comment]
> > Try to find a workaround, don't wait anything from Cfx, it been many years like that.
> 
> The only workaround is to reduce the number of .ymt files, which is clearly not feasible due to certain assets being locked in escrow. For example, ONX faces already use up 2 .ymts — one for male and one for female. Then, the Emergency Uniform Pack takes up another 2 .ymts. That only leaves a single slot available for the latest Money Fronts build, given the 5 .ymt file limit.
> 
> > > With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, [@DaniGP17](https://github.com/DaniGP17)'s pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.
> > 
> > 
> > The pool size increases that were made are much easier than what's being requested here. It's difficult to raise this limit without causing other problems. Even on the latest game build you can already have hundreds of clothing items, you don't need to have a different ymt for every single clothing piece you purchase. You just need the clothing within that resource but you can place them all inside a single ymt.
> 
> So you're suggesting we merge all the .ymt files into one? Isn't there a limit to how many clothing items can be included per pack — like 128 per clothing category in a pack, for example? Plus, doesn't asset escrow come into play here? I don't think merging is a realistic option, especially when people are asset-escrowing their clothing packs.

I know with my server, using YMTEditor I have combined both my escrowed and non-escrowed clothing packs into one .YMT file, as long as the original folders exist for the escrowed clothing you can just rename them to follow the .YMT

I have 1 .YMT for Police, 1 for Ambo, 1 for Fire, 1 for Civilian clothes and it all works.

However, I'm not close to hitting high clothing numbers per YMT file.

**Comment by SpellLivedBackwards** [comment]
> > > Try to find a workaround, don't wait anything from Cfx, it been many years like that.
> > 
> > 
> > The only workaround is to reduce the number of .ymt files, which is clearly not feasible due to certain assets being locked in escrow. For example, ONX faces already use up 2 .ymts — one for male and one for female. Then, the Emergency Uniform Pack takes up another 2 .ymts. That only leaves a single slot available for the latest Money Fronts build, given the 5 .ymt file limit.
> > > > With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, [@DaniGP17](https://github.com/DaniGP17)'s pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.
> > > 
> > > 
> > > The pool size increases that were made are much easier than what's being requested here. It's difficult to raise this limit without causing other problems. Even on the latest game build you can already have hundreds of clothing items, you don't need to have a different ymt for every single clothing piece you purchase. You just need the clothing within that resource but you can place them all inside a single ymt.
> > 
> > 
> > So you're suggesting we merge all the .ymt files into one? Isn't there a limit to how many clothing items can be included per pack — like 128 per clothing category in a pack, for example? Plus, doesn't asset escrow come into play here? I don't think merging is a realistic option, especially when people are asset-escrowing their clothing packs.
> 
> I know with my server, using YMTEditor I have combined both my escrowed and non-escrowed clothing packs into one .YMT file, as long as the original folders exist for the escrowed clothing you can just rename them to follow the .YMT
> 
> I have 1 .YMT for Police, 1 for Ambo, 1 for Fire, 1 for Civilian clothes and it all works.
> 
> However, I'm not close to hitting high clothing numbers per YMT file.

If your on the latest game version, you only get 5 `.ymts`. So, you'd hit if you added a new one.

**Comment by Scullyy** [comment]
> > Try to find a workaround, don't wait anything from Cfx, it been many years like that.
> 
> The only workaround is to reduce the number of .ymt files, which is clearly not feasible due to certain assets being locked in escrow. For example, ONX faces already use up 2 .ymts — one for male and one for female. Then, the Emergency Uniform Pack takes up another 2 .ymts. That only leaves a single slot available for the latest Money Fronts build, given the 5 .ymt file limit.
> 
> > > With all due respect, what frustrates many people on the platform is the lack of communication from the team. On June 10th, [@DaniGP17](https://github.com/DaniGP17)'s pull request was closed with the explanation that the team was working on a solution. It's now been nearly three months with no follow-up. Meanwhile, requests for pool size increases are handled within just a few days. I don’t understand the lack of communication or updates on this issue. It’s preventing many servers from upgrading to the latest downloadable content, including vehicles, weapons, and more. What’s even more confusing is that a specific build is now being recommended, yet there’s still no solution in place to address this problem.
> > 
> > 
> > The pool size increases that were made are much easier than what's being requested here. It's difficult to raise this limit without causing other problems. Even on the latest game build you can already have hundreds of clothing items, you don't need to have a different ymt for every single clothing piece you purchase. You just need the clothing within that resource but you can place them all inside a single ymt.
> 
> So you're suggesting we merge all the .ymt files into one? Isn't there a limit to how many clothing items can be included per pack — like 128 per clothing category in a pack, for example? Plus, doesn't asset escrow come into play here? I don't think merging is a realistic option, especially when people are asset-escrowing their clothing packs.

128 per category sounds about right which would leave you with a total of over 1k clothing items per ymt. You could also edit the ymt files that ship with the game and put them on your server as you can still add a lot of slots to those but in the latest game build you have 6 free ymt slots so that should be more than enough anyways. Escrow doesn't come into play here as the clothing doesn't need to be in the same resource as the ymt.

**Comment by SpellLivedBackwards** [comment]
> 128 per category sounds about right which would leave you with a total of over 1k clothing items per ymt. You could also edit the ymt files that ship with the game and put them on your server as you can still add a lot of slots to those but in the latest game build you have 6 free ymt slots so that should be more than enough anyways. Escrow doesn't come into play here as the clothing doesn't need to be in the same resource as the ymt.

**In latest game version b3570 (mp2025_01), we have about 5 additional free .ymt slots to be used**. Which is the latest DLC.

**Comment by Scullyy** [comment]
> **In latest game version b3570 (mp2025_01), we have about 5 additional free .ymt slots to be used**. Which is the latest DLC.

My mistake, this is still thousands of clothing slots though, should be more than enough. There isn't really a need for this to be rushed.

**Comment by Gogsi** [comment]
It's 128 drawables per type, per gender. Each drawable can have 26 variations/retextures. That makes it 3328 variations of a clothing type (say shirt) for each gender in a single ymt. Or with with 5 slots (on 3570) and 2 genders, you can have 33 280 "different" shirts (although only 1280 unique models). 

There are 12 clothing types which brings the theoretical maximum to a total of 399 360 clothing items, almost 400k. Of course that will be hard to reach as you would have to make 33k different hairstyles and accessories, etc.

I'm not sure if any of the community made clothing tools can automate splitting all of your clothing resources into fully packed ymts. Seems like a really nice tool to have, as the slots themselves are actually plenty, it's just hard to fill them well.

**Comment by SpellLivedBackwards** [comment]
> It's 128 drawables per type, per gender. Each drawable can have 26 variations/retextures. That makes it 3328 variations of a clothing type (say shirt) for each gender in a single ymt. Or with with 5 slots (on 3570) and 2 genders, you can have 33 280 "different" shirts (although only 1280 unique models).
> 
> There are 12 clothing types which brings the theoretical maximum to a total of 399 360 clothing items, almost 400k. Of course that will be hard to reach as you would have to make 33k different hairstyles and accessories, etc.
> 
> I'm not sure if any of the community made clothing tools can automate splitting all of your clothing resources into fully packed ymts. Seems like a really nice tool to have, as the slots themselves are actually plenty, it's just hard to fill them well.

There's tools but what will do the automation for the encrypted resources?

**Comment by Gogsi** [comment]
The ymt file itself is not encrypted. If the tool can handle the model being encrypted without crashing, it should be able to fill out the ymt slots.

**Comment by Scullyy** [comment]
> The ymt file itself is not encrypted. If the tool can handle the model being encrypted without crashing, it should be able to fill out the ymt slots.

Wouldn’t be necessary to handle the model under any circumstance either tbh, file would be generated by reading the component type from the model file name. It wouldn’t need to open it or read any encrypted data.

**Comment by Gogsi** [comment]
> Wouldn’t be necessary to handle the model under any circumstance either tbh, file would be generated by reading the component type from the model file name. It wouldn’t need to open it or read any encrypted data.

It can be done for sure. I just feel like existing tools would try to read the model itself for visualization reasons and error out. But it should be tested.

**Comment by SpellLivedBackwards** [comment]
> > Wouldn’t be necessary to handle the model under any circumstance either tbh, file would be generated by reading the component type from the model file name. It wouldn’t need to open it or read any encrypted data.
> 
> It can be done for sure. I just feel like existing tools would try to read the model itself for visualization reasons and error out. But it should be tested.

Tedious, don't see why it can't be fixed 🤷

**Comment by Scullyy** [comment]
> > > Wouldn’t be necessary to handle the model under any circumstance either tbh, file would be generated by reading the component type from the model file name. It wouldn’t need to open it or read any encrypted data.
> > 
> > 
> > It can be done for sure. I just feel like existing tools would try to read the model itself for visualization reasons and error out. But it should be tested.
> 
> Tedious, don't see why it can't be fixed 🤷

If you don’t see why it can’t be done then you must know how it can be done considering you don’t see any obstacles. You’re free to make a PR lol. On a serious note, it can be done but it’s quite difficult without causing some other issue.

**Comment by CodexisPhantom** [comment]
Hey guys,
Could someone explain the issue with YMTs to me? I get the general problems with clothes, but why exactly can’t this be done? And if I wanted to start working on a fix, what would be the right approach?

**Comment by SpellLivedBackwards** [comment]
> Hey guys, Could someone explain the issue with YMTs to me? I get the general problems with clothes, but why exactly can’t this be done? And if I wanted to start working on a fix, what would be the right approach?

Probably @DaniGP17 knows more but he made a request here: https://github.com/citizenfx/fivem/pull/3444

**Comment by DaniGP17** [comment]
> Hey guys, Could someone explain the issue with YMTs to me? I get the general problems with clothes, but why exactly can’t this be done? And if I wanted to start working on a fix, what would be the right approach?

I'm going to explain the current problem as I understood it in my research and what I remember: currently there is a global constant that is the maximum amount of dependencies per stream, in the case of builds>= 2612 it is 100, but in previous ones it is 80, this constant is being used in quite a few places including for other types of streams not just ymt, I remember that it was also used in something related to vehicles, increasing this limit is quite complicated because it consists of many arrays located in the stack of the functions and therefore it is necessary to increase the size reserved in the stack, the accesses to those arrays and the variables that are located after. In each build the maximum number is reduced because more ymts are added to the base game. I recommend that you look at the PR(#3444) to see how I started trying to increase the limit, I have not continued because it caused a crash when changing peds, but I think it was caused because when increasing the size of an atArray I was not taking into account that the m_count was located at the end and other functions believed that the m_count was still at an offset when when increasing the array it should have moved.

**Comment by CodexisPhantom** [comment]
> > Hey guys, Could someone explain the issue with YMTs to me? I get the general problems with clothes, but why exactly can’t this be done? And if I wanted to start working on a fix, what would be the right approach?
> 
> I'm going to explain the current problem as I understood it in my research and what I remember: currently there is a global constant that is the maximum amount of dependencies per stream, in the case of builds>= 2612 it is 100, but in previous ones it is 80, this constant is being used in quite a few places including for other types of streams not just ymt, I remember that it was also used in something related to vehicles, increasing this limit is quite complicated because it consists of many arrays located in the stack of the functions and therefore it is necessary to increase the size reserved in the stack, the accesses to those arrays and the variables that are located after. In each build the maximum number is reduced because more ymts are added to the base game. I recommend that you look at the PR([#3444](https://github.com/citizenfx/fivem/pull/3444)) to see how I started trying to increase the limit, I have not continued because it caused a crash when changing peds, but I think it was caused because when increasing the size of an atArray I was not taking into account that the m_count was located at the end and other functions believed that the m_count was still at an offset when when increasing the array it should have moved.

Okey I understand the problem now, thanks for explaining.

**Comment by dlikedior** [comment]
> In the latest game version **(b3717, mp2025_02)**, there are now approximately 4 additional free .ymt slots available.

This issue has been open for over a year, yet players still can’t upgrade to the latest build without reducing the size of their clothing packs. For instance, if you're using Onx Faces and their Emergency Uniform Pack, that already takes up 4 slots. Why isn't this a priority, especially when most server owners want to upgrade to the latest game build?

**Comment by Legacy-TacticalGamingInteractive** [comment]
Indeed. I wish. Maybe it's too big of a project and they don't have time for it. Just like increasing 255 limit for carcols from 8 bit to 16 bit integer...

Plus all these EUP devs have everything escrowed of course so you can't really buy any more EUP and put it in easily like that. Meaning they are losing income from Tebex.

**Comment by dlikedior** [comment]
> Indeed. I wish. Maybe it's too big of a project and they don't have time for it. Just like increasing 255 limit for carcols from 8 bit to 16 bit integer...
> 
> Plus all these EUP devs have everything escrowed of course so you can't really buy any more EUP and put it in easily like that. Meaning they are losing income from Tebex.

It’s unfortunate that we haven’t received any updates from the team, except from someone who contributed and then closed their request. @DaniGP17, since you communicate with us more frequently than they do, do you have any news on this?

**Comment by DaniGP17** [comment]
> @DaniGP17, since you communicate with us more frequently than they do, do you have any news on this?

I'm just a contributor but very sure they are working in this limit and other ones like max portals.

**Comment by n3xuuu** [comment]
> I'm just a contributor but very sure they are working in this limit and other ones like max portals.

One of the elements mentioned in a pull request that they are in a ‘feature freeze’ for FiveM. This might be because the new platform is closer to release, but I can’t give an exact answer ;/

**Comment by dlikedior** [comment]
Once again, there’s a lack of communication from the team. No server can update to the latest version of the game because of these limitations. Even with the feature freeze, this issue has persisted through almost every version update. Most servers want to use the new content but can’t due to the ymt clothing restrictions.

**Comment by lLukDreams** [comment]
from what i know its not easy to support higher limits, probably is the new platform so close it would not make sense to start with it since by the time its done, everyone is mgirating to the new one or whatever, but you are right, communication is really $h1t

**Comment by Gogsi** [comment]
Increasing the limit is tough, because over 50 functions that allocate arrays with the size of 100 (the maximum number of ymts) have to patched one by one. Every such patch will decrease performance a bit too, and I think in general having more ymts always loaded may have a noticeable impact on streaming other files (such as the map).

I think a simpler solution would be for a guide or tool to be made that allows you to pack Rockstar's DLC clothing into as few ymt files as possible. Currently their approach is very inefficient, because they use a whole ymt slot for just the dozen or so clothes they add per DLC, when it could theoretically hold hundreds. Packing them as densely as possible, will allow servers to stream some of their clothes as replacement, instead of addon ymts, allowing for more overall packs on the server. I think I will work a bit on that over the holidays and see if I can come up with an efficient way to do it. In theory Escrow shouldn't be a concern for this either, because the ymt files themselves aren't Escrowed.

**Comment by Legacy-TacticalGamingInteractive** [comment]
Not a bad idea

> Increasing the limit is tough, because over 50 functions that allocate arrays with the size of 100 (the maximum number of ymts) have to patched one by one. Every such patch will decrease performance a bit too, and I think in general having more ymts always loaded may have a noticeable impact on streaming other files (such as the map).
> 
> I think a simpler solution would be for a guide or tool to be made that allows you to pack Rockstar's DLC clothing into as few ymt files as possible. Currently their approach is very inefficient, because they use a whole ymt slot for just the dozen or so clothes they add per DLC, when it could theoretically hold hundreds. Packing them as densely as possible, will allow servers to stream some of their clothes as replacement, instead of addon ymts, allowing for more overall packs on the server. I think I will work a bit on that over the holidays and see if I can come up with an efficient way to do it. In theory Escrow shouldn't be a concern for this either, because the ymt files themselves aren't Escrowed.

**Comment by Scullyy** [comment]
4 ymt slots are available in this DLC, using one for emergency uniforms such as PD and EMS and using another for CIVs you should be perfectly fine. Many servers don't wish to take the time to merge them though. A tool to merge streamed ymts and rename the files automatically would likely be better than merging all of the vanilla files. I'm not sure why CFX can't just ask Rockstar to raise such limit again anyways, they raised the limit before. I'm sure they can push a simple patch to free up 10 slots max reserved for FiveM.

**Comment by dlikedior** [comment]
> 4 ymt slots are available in this DLC, using one for emergency uniforms such as PD and EMS and using another for CIVs you should be perfectly fine. Many servers don't wish to take the time to merge them though. A tool to merge streamed ymts and rename the files automatically would likely be better than merging all of the vanilla files. I'm not sure why CFX can't just ask Rockstar to raise such limit again anyways, they raised the limit before. I'm sure they can push a simple patch to free up 10 slots max reserved for FiveM.

Regardless of whether you believe "4 YMTS are enough", you're mistaken. Most servers have more than 5, and many assets are now escrowed, especially when it comes to clothing. You can't add clothing for civilians or anything else. I don't understand why the Cfx.re team hasn't addressed this issue, as the limitations have been present for years. Many server owners can't upgrade to the latest DLC because of it. Most server owners want to be on the latest game build, so either Rockstar should address this or at least provide a response by the Cfx.re Team. I don't see why they can't do this due to the "new platform" restrictions.

**Comment by ghost** [comment]
Even with escrowed assets, you can still set up a YMT file that allows both non-escrowed and escrowed clothing to work.

I've used DurtyClothTool, added both in and it's renamed the files.

All you have to make sure if the escrowed files go back into the resource folder they came from originally.

**Comment by dlikedior** [comment]
> Even with escrowed assets, you can still set up a YMT file that allows both non-escrowed and escrowed clothing to work.
> 
> I've used DurtyClothTool, added both in and it's renamed the files.
> 
> All you have to make sure if the escrowed files go back into the resource folder they came from originally.

Can you merge all .ymts into 1 huge .ymt?

**Comment by ghost** [comment]
> > Even with escrowed assets, you can still set up a YMT file that allows both non-escrowed and escrowed clothing to work.
> > I've used DurtyClothTool, added both in and it's renamed the files.
> > All you have to make sure if the escrowed files go back into the resource folder they came from originally.
> 
> Can you merge all .ymts into 1 huge .ymt?

I'm sure there's a way (eg opening the YMT's if they themselves have not been escrowed, exporting to XML and combining that way) but what I have found is to just use DurtyClothTool or any other tool that allows you to combine EUP and do it that way, it exports everything you need including the creaturemetadata and the YMT's and works fine.

Obviously everyone is going to have different use cases but that's what has been working for me.

**Comment by dlikedior** [comment]
> > > Even with escrowed assets, you can still set up a YMT file that allows both non-escrowed and escrowed clothing to work.
> > > I've used DurtyClothTool, added both in and it's renamed the files.
> > > All you have to make sure if the escrowed files go back into the resource folder they came from originally.
> > 
> > 
> > Can you merge all .ymts into 1 huge .ymt?
> 
> I'm sure there's a way (eg opening the YMT's if they themselves have not been escrowed, exporting to XML and combining that way) but what I have found is to just use DurtyClothTool or any other tool that allows you to combine EUP and do it that way, it exports everything you need including the creaturemetadata and the YMT's and works fine.
> 
> Obviously everyone is going to have different use cases but that's what has been working for me.

The issue is that the escrowed ones are the problem; even so, DurtyClothingTool generates more than .ymt.

**Comment by Scullyy** [comment]
> > 4 ymt slots are available in this DLC, using one for emergency uniforms such as PD and EMS and using another for CIVs you should be perfectly fine. Many servers don't wish to take the time to merge them though. A tool to merge streamed ymts and rename the files automatically would likely be better than merging all of the vanilla files. I'm not sure why CFX can't just ask Rockstar to raise such limit again anyways, they raised the limit before. I'm sure they can push a simple patch to free up 10 slots max reserved for FiveM.
> 
> Regardless of whether you believe "4 YMTS are enough", you're mistaken. Most servers have more than 5, and many assets are now escrowed, especially when it comes to clothing. You can't add clothing for civilians or anything else. I don't understand why the Cfx.re team hasn't addressed this issue, as the limitations have been present for years. Many server owners can't upgrade to the latest DLC because of it. Most server owners want to be on the latest game build, so either Rockstar should address this or at least provide a response by the Cfx.re Team. I don't see why they can't do this due to the "new platform" restrictions.

Okay so you have 5+ so merge them so you only have 4 or less then. You can do it with escrowed assets too. This has never really been an issue, you just likely never knew how to merge them. Not knowing how or not wanting to take the time to do something doesn’t mean it’s an issue.

**Comment by alberttheprince** [comment]
> you have 5+ so merge them so you only have 4 or less then. You can do it with escrowed assets too. This has never really been an issue, you just likely never knew how to merge them. Not knowing how or not wanting to take the time to do something doesn’t mean it’s an issue.

"Just make big number into small number"

I don't know why I'm bothering to reply to bad faith response, but no, it's not that simple. 

High heels and hats also create a seperate creaturemetadata file, which counts against this limit. So you'd need to condense 5 DLCs into 3 if you have any addon hats or high heels as they count as a ymt, which also a big ask. 

You are not magically going to turn 640 jbib clothing items into 384 clothing items, sure we can all just agree to delete massive amounts of clothing from packs, but it's a important feature of FiveM, and a big roadblock for a lot of servers moving to newer game versions. It's literally a paid feature for server owners, so it's not even like we're asking for a "free" feature to be worked on.

**Comment by VexoaXYZ** [comment]
I've built a tool to solve this problem. It's completely free.

Check it out here: [Merge That YMT](https://cfx.software/clothing-merger)

**Comment by Erumeldor** [comment]
This issue needs to be pushed.

With the December 2025 update, the limit has already been reduced again, and we are now at 4 YMTs per gender.

In practice, this only allows about two usable packs, depending on how they are designed. For example, if both include heels, that already uses up the available space.

Since custom clothing streaming is a paid feature, this restriction is becoming quite limiting.

There is also a concern that a future update may reduce this limit even further, if another update is released.

**Comment by Legacy-TacticalGamingInteractive** [comment]
Agreed. We are basically screwed if we want to have any EUP. Im stuck on 3258...

Doesnt help that so many devs use Tebex for their clothing and its encypted so youre not able to even merge it probably into one pack.

So now its at the point where its also hurting Tebex sales and revenue for CFX.

**Comment by Scullyy** [comment]
> This issue needs to be pushed.
> 
> With the December 2025 update, the limit has already been reduced again, and we are now at 4 YMTs per gender.
> 
> In practice, this only allows about two usable packs, depending on how they are designed. For example, if both include heels, that already uses up the available space.
> 
> Since custom clothing streaming is a paid feature, this restriction is becoming quite limiting.
> 
> There is also a concern that a future update may reduce this limit even further, if another update is released.

It’s best to wait until the next update is actually out to see if it’s even necessary as they could increase the limit in the update too. There is also a feature freeze as the team is working on an updated version of FiveM which is closed source iirc so even if there isn’t an increase you’ll just have to deal with it for now. Right now 4 is enough, just merge them.

**Comment by Scullyy** [comment]
> Agreed. We are basically screwed if we want to have any EUP. Im stuck on 3258...
> 
> Doesnt help that so many devs use Tebex for their clothing and its encypted so youre not able to even merge it probably into one pack.
> 
> So now its at the point where its also hurting Tebex sales and revenue for CFX.

Nothing is stopping you from merging your clothing, you don’t need the streamed files to be in the same resource as the ymt. Escrow doesn’t limit you at all.

**Comment by Erumeldor** [comment]
> It’s best to wait until the next update is actually out to see if it’s even necessary as they could increase the limit in the update too. 

Agreed. But since the last time the limit was raised while the free ymt limit was <2, and this time there won't be unlimited GTA Online updates anymore with GTA 6 in the go, I doubt that raise.



> ... so even if there isn’t an increase you’ll just have to deal with it for now. Right now 4 is enough, just merge them.


> Nothing is stopping you from merging your clothing, you don’t need the streamed files to be in the same resource as the ymt. Escrow doesn’t limit you at all.

Stupid thing to say, if you don't know the circumstances.

With 4 ymt free, one lost for heels, you have 3 per gender. This is not enough for all our current clothes, not speaking of adding new ones for the community.
We are therefore one update behind, to have at least 5 ymts.

It's not about merging, it's about the total amount that's limiting.

**Comment by Scullyy** [comment]
> > It’s best to wait until the next update is actually out to see if it’s even necessary as they could increase the limit in the update too.
> 
> Agreed. But since the last time the limit was raised while the free ymt limit was <2, and this time there won't be unlimited GTA Online updates anymore with GTA 6 in the go, I doubt that raise.
> 
> > ... so even if there isn’t an increase you’ll just have to deal with it for now. Right now 4 is enough, just merge them.
> 
> > Nothing is stopping you from merging your clothing, you don’t need the streamed files to be in the same resource as the ymt. Escrow doesn’t limit you at all.
> 
> Stupid thing to say, if you don't know the circumstances.
> 
> With 4 ymt free, one lost for heels, you have 3 per gender. This is not enough for all our current clothes, not speaking of adding new ones for the community. We are therefore one update behind, to have at least 5 ymts.
> 
> It's not about merging, it's about the total amount that's limiting.

As Gogsi mentioned above, you can have 128 items for each type for both male and female peds. Each of those items also have 26 texture slots which is 3328 total items per clothing type. You can also go back and edit the official dlcs to add even more slots as some of them are practically empty. It’s very rude to state what I said is stupid when you’re the one who is incapable of simply merging your clothing just because it’s “time consuming” or you lack the knowledge to do it. Raising the limit isn’t easy, every proposal so far has lead to other issues. I think you’ll do fine on the latest DLC considering you can have thousands of clothes. I’m sure you also have multiple clothing items using a copy of the same model with different textures, remove the duplicates and put them under a single item making use of the 26 texture slots.

**Comment by Legacy-TacticalGamingInteractive** [comment]
128 items lol.

Considering the two biggest types are jackets and shirts... that extremely quickly gets maxed out at just 128 drawables per merged pack per gender. Escrowed or not. Hell even 256 is nothing if you have two merged packs and 4-6 ymt. Some LEO departments have 30 jackets alone... or shirt drawables a piece. You can have thousands of clothes but thats if you are counting drawable types that arent even that popularly used. The real issue boils down to a lack of availability for jackets, shirts, hats/helmets and perhaps shoes. 128 per pack is nothing... so while you try to make it sound like we have thousands of available clothing slots... missing the very important detail and distinction of drawable types... its rather disingenuous.

**Comment by alberttheprince** [comment]
> As Gogsi mentioned above, you can have 128 items for each type for both male and female peds. Each of those items also have 26 texture slots which is 3328 total items per clothing type. 

The problem here is that you're just oversimplying things because you're either ignorant or unable to admit to being wrong. Considering that clothing and accessories are seperate, and on top of that clothing with physics and other special properties can't just go in any category, AND high heels and other special clothing, you're just wasting everyone's time by inserting yourself in a discussion that you have no place being in.

>You can also go back and edit the official dlcs to add even more slots as some of them are practically empty.

The problem with this is we are clearly looking to keep the base game assets, otherwise we'd all be doing that.

> It’s very rude to state 

I'd consider your weird behaviour in this discussion to be rude, as it really benefits no one and jusrt serves to waste everyone's time. We don't need lessons from someone who barely grasps the topic of clothing and addons. We know we can just jam random clothing assets into other categories that they don't belong in and replace base game clothing.

>  Raising the limit isn’t easy

Why do you care if it's easy or not? You don't work for CFX or contribute in any meaningful way to the project, so leave it to them to respond if it's easy or not. CFX doesn't need volunteers to complain about it.

**Comment by Legacy-TacticalGamingInteractive** [comment]
Come on guys lets get back on topic shall we?

**Comment by xalva98** [comment]
> > As Gogsi mentioned above, you can have 128 items for each type for both male and female peds. Each of those items also have 26 texture slots which is 3328 total items per clothing type.
> 
> The problem here is that you're just oversimplying things because you're either ignorant or unable to admit to being wrong. Considering that clothing and accessories are seperate, and on top of that clothing with physics and other special properties can't just go in any category, AND high heels and other special clothing, you're just wasting everyone's time by inserting yourself in a discussion that you have no place being in.
> 
> > You can also go back and edit the official dlcs to add even more slots as some of them are practically empty.
> 
> The problem with this is we are clearly looking to keep the base game assets, otherwise we'd all be doing that.
> 
> > It’s very rude to state
> 
> I'd consider your weird behaviour in this discussion to be rude, as it really benefits no one and jusrt serves to waste everyone's time. We don't need lessons from someone who barely grasps the topic of clothing and addons. We know we can just jam random clothing assets into other categories that they don't belong in and replace base game clothing.
> 
> > Raising the limit isn’t easy
> 
> Why do you care if it's easy or not? You don't work for CFX or contribute in any meaningful way to the project, so leave it to them to respond if it's easy or not. CFX doesn't need volunteers to complain about it.

Just wait until enhanced hits I'm fairly certain this will be a non issue in the future.

**Comment by Scullyy** [comment]
> 128 items lol.
> 
> Considering the two biggest types are jackets and shirts... that extremely quickly gets maxed out at just 128 drawables per merged pack per gender. Escrowed or not. Hell even 256 is nothing if you have two merged packs and 4-6 ymt. Some LEO departments have 30 jackets alone... or shirt drawables a piece. You can have thousands of clothes but thats if you are counting drawable types that arent even that popularly used. The real issue boils down to a lack of availability for jackets, shirts, hats/helmets and perhaps shoes. 128 per pack is nothing... so while you try to make it sound like we have thousands of available clothing slots... missing the very important detail and distinction of drawable types... its rather disingenuous.

I really don’t see how 256 models isn’t enough per type, surely you can’t be counting unique models. Majority of custom clothing are usually re-textures of vanilla clothes, you could reduce the number greatly by adding additional texture slots to the original vanilla slot for that model and for your LEO you can use the same slots with different textures for majority of those. You’ll have to work within the limit as there is a feature freeze regardless.

**Comment by Scullyy** [comment]
> > As Gogsi mentioned above, you can have 128 items for each type for both male and female peds. Each of those items also have 26 texture slots which is 3328 total items per clothing type.
> 
> The problem here is that you're just oversimplying things because you're either ignorant or unable to admit to being wrong. Considering that clothing and accessories are seperate, and on top of that clothing with physics and other special properties can't just go in any category, AND high heels and other special clothing, you're just wasting everyone's time by inserting yourself in a discussion that you have no place being in.
> 
> > You can also go back and edit the official dlcs to add even more slots as some of them are practically empty.
> 
> The problem with this is we are clearly looking to keep the base game assets, otherwise we'd all be doing that.
> 
> > It’s very rude to state
> 
> I'd consider your weird behaviour in this discussion to be rude, as it really benefits no one and jusrt serves to waste everyone's time. We don't need lessons from someone who barely grasps the topic of clothing and addons. We know we can just jam random clothing assets into other categories that they don't belong in and replace base game clothing.
> 
> > Raising the limit isn’t easy
> 
> Why do you care if it's easy or not? You don't work for CFX or contribute in any meaningful way to the project, so leave it to them to respond if it's easy or not. CFX doesn't need volunteers to complain about it.

Clearly you’re upset based on the very toxic response you just provided but you seem to be confused. There is currently a feature freeze so the only productive help you can currently provide here is informing people how to work within the limit. You also seem to think I said you should replace vanilla clothing when I never suggested that at all, based on the emotions in your response you likely rambled away before understanding what I said but to be clear you don’t need to replace vanilla clothing, there is thousands of leftover slots left in the vanilla DLCs which you can modify to max them all out. This would be equal to adding at least 4 more ymt slots, my estimation is probably below the actual number though.

**Comment by st860923** [comment]
I've been following this issue since FEB. and have been working on a potential solution that avoids the need to patch hardcoded stack sizes, which can be brittle and version specific.
My approach involves hooking the GetDependencies function in the archetype and metadata vtables. Instead of trying to increase a fixed size pool, I chain any excess metadata dependencies into a linked list.
When the game requests dependencies, the recursive expansion process naturally walks this chain. This means each individual call to GetDependencies only returns a small, manageable number of entries, completely avoiding the overflow that causes crashes. The key benefit is that there's no hard limit on the total number of dependencies.
I've got a POC working on build 3717. There are some valid considerations, such as the performance impact and ensuring that all callers of GetDependencies correctly handle the recursion. However, my testing indicates that the critical game functions that handle asset loading already recurse, so they process the full chain correctly. The main performance bottleneck in model loading remains disk I/O and texture decompression.
This "chaining" method seems to be a more robust and future proof way to solve the .ymt limit problem for clothing and other assets.

<img width="1591" height="897" alt="Image" src="https://github.com/user-attachments/assets/0db3da0f-7aa9-414f-aeb6-1c8dae601606" />

**Comment by CodexisPhantom** [comment]
> I've been following this issue since FEB. and have been working on a potential solution that avoids the need to patch hardcoded stack sizes, which can be brittle and version specific. My approach involves hooking the GetDependencies function in the archetype and metadata vtables. Instead of trying to increase a fixed size pool, I chain any excess metadata dependencies into a linked list. When the game requests dependencies, the recursive expansion process naturally walks this chain. This means each individual call to GetDependencies only returns a small, manageable number of entries, completely avoiding the overflow that causes crashes. The key benefit is that there's no hard limit on the total number of dependencies. I've got a POC working on build 3717. There are some valid considerations, such as the performance impact and ensuring that all callers of GetDependencies correctly handle the recursion. However, my testing indicates that the critical game functions that handle asset loading already recurse, so they process the full chain correctly. The main performance bottleneck in model loading remains disk I/O and texture decompression. This "chaining" method seems to be a more robust and future proof way to solve the .ymt limit problem for clothing and other assets.
> <img alt="Image" width="1591" height="897" src="https://private-user-images.githubusercontent.com/68638808/589889443-0db3da0f-7aa9-414f-aeb6-1c8dae601606.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzgzMTIwOTQsIm5iZiI6MTc3ODMxMTc5NCwicGF0aCI6Ii82ODYzODgwOC81ODk4ODk0NDMtMGRiM2RhMGYtN2FhOS00MTRmLWFlYjYtMWM4ZGFlNjAxNjA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTA5VDA3Mjk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2NmViZjZiOWRhYmUyOWI1MjFmMmFhOGMxMzM4NWRiZDVmOTQ5ODhkOTEzODgwODY4MTRhZGMyZDMwYTdiOTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.MMDbedR4P1dmPK_tILK_o8sukqIrcfO66lYfVatoeCc">

If you need help contact me on discord : codexis

**Comment by 4azm** [comment]
> I've been following this issue since FEB. and have been working on a potential solution that avoids the need to patch hardcoded stack sizes, which can be brittle and version specific. My approach involves hooking the GetDependencies function in the archetype and metadata vtables. Instead of trying to increase a fixed size pool, I chain any excess metadata dependencies into a linked list. When the game requests dependencies, the recursive expansion process naturally walks this chain. This means each individual call to GetDependencies only returns a small, manageable number of entries, completely avoiding the overflow that causes crashes. The key benefit is that there's no hard limit on the total number of dependencies. I've got a POC working on build 3717. There are some valid considerations, such as the performance impact and ensuring that all callers of GetDependencies correctly handle the recursion. However, my testing indicates that the critical game functions that handle asset loading already recurse, so they process the full chain correctly. The main performance bottleneck in model loading remains disk I/O and texture decompression. This "chaining" method seems to be a more robust and future proof way to solve the .ymt limit problem for clothing and other assets.
> 
> <img alt="Image" width="1591" height="897" src="https://private-user-images.githubusercontent.com/68638808/589889443-0db3da0f-7aa9-414f-aeb6-1c8dae601606.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA5NDYwOTYsIm5iZiI6MTc4MDk0NTc5NiwicGF0aCI6Ii82ODYzODgwOC81ODk4ODk0NDMtMGRiM2RhMGYtN2FhOS00MTRmLWFlYjYtMWM4ZGFlNjAxNjA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjA4VDE5MDk1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiZWJlMDMzMjJlZjNlY2QwNWIxYjVmYWMwZGRjOGRhNjFiMTk1MWYxNGNkOGE1NDRkYTNiNjFiZDFiZWRkMzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.KLfGC9VNYVrycTV9I4oUyQ9u8emk3_fvO1C6NSQC0yY">

Are you still working on this?

**Comment by DaniGP17** [comment]
There's a feature branch for a possible fix, if  someone want to give a try: `specific/feature/experimental-ymt-extension`
